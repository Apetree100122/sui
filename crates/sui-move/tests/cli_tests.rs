// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use insta_cmd::{assert_cmd_snapshot, get_cargo_bin};
use std::fs;
use std::path::Path;
use std::process::Command;
use walkdir::WalkDir;

// [test_shell_snapshot] is run on every file matching [TEST_PATTERN] in [TEST_DIR]; this runs the
// files as shell scripts and compares their output to the snapshots; use `cargo insta test
// --review` to update the snapshots.

const TEST_DIR: &str = "tests/tests";
const TEST_PATTERN: &str = r"^test.*\.sh$";

/// run the bash script at [path], comparing its output to the insta snapshot of the same name.
/// The script is run in a temporary working directory that contains a copy of the parent directory
/// of [path], with the `sui-move` binary on the path.
fn test_shell_snapshot(path: &Path) -> datatest_stable::Result<()> {
    // copy files into temporary directory
    let srcdir = path.parent().unwrap();
    let tmpdir = tempfile::tempdir()?;
    let sandbox = tmpdir.path().join("sandbox");

    for entry in WalkDir::new(srcdir) {
        let entry = entry.unwrap();
        let srcfile = entry.path();
        let dstfile = sandbox.join(srcfile.strip_prefix(srcdir)?);
        if srcfile.is_dir() {
            fs::create_dir_all(dstfile)?;
        } else {
            fs::copy(srcfile, dstfile)?;
        }
    }

    // set up path
    // Note: we need to create a symlink instead of just adding the bin dir to the path to prevent
    // local pathnames from leaking into the snapshot files.
    std::os::unix::fs::symlink(
        get_cargo_bin("sui-move").parent().unwrap(),
        tmpdir.path().join("bin"),
    )?;

    // set up command
    let mut shell = Command::new("bash");
    shell
        .env("PATH", "/bin:/usr/bin:../bin")
        .current_dir(sandbox)
        .arg(path.file_name().unwrap());

    // run it!
    let snapshot_name: String = path
        .strip_prefix("tests/tests")?
        .to_string_lossy()
        .to_string();

    assert_cmd_snapshot!(snapshot_name, shell);

    Ok(())
}

datatest_stable::harness!(test_shell_snapshot, TEST_DIR, TEST_PATTERN);
