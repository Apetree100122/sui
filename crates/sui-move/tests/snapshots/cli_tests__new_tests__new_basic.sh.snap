---
source: crates/sui-move/tests/cli_tests.rs
info:
  program: bash
  args:
    - new_basic.sh
  env:
    PATH: "/bin:/usr/bin:../bin"
---
success: true
exit_code: 0
----- stdout -----
==== files in project ====
Move.toml
sources
tests
==== files in sources ====
myproj.move
==== files in tests =====
myproj_tests.move
==== myproj/Move.toml ====
[package]
name = "myproj"
edition = "2024.beta" # edition = "legacy" to use legacy (pre-2024) Move
# license = ""           # e.g., "MIT", "GPL", "Apache 2.0"
# authors = ["..."]      # e.g., ["Joe Smith (joesmith@noemail.com)", "John Snow (johnsnow@noemail.com)"]

[dependencies]
Sui = { git = "https://github.com/MystenLabs/sui.git", subdir = "crates/sui-framework/packages/sui-framework", rev = "framework/testnet" }

# For remote import, use the `{ git = "...", subdir = "...", rev = "..." }`.
# Revision can be a branch, a tag, and a commit hash.
# MyRemotePackage = { git = "https://some.remote/host.git", subdir = "remote/path", rev = "main" }

# For local dependencies use `local = path`. Path is relative to the package root
# Local = { local = "../path/to" }

# To resolve a version conflict and force a specific version for dependency
# override use `override = true`
# Override = { local = "../conflicting/version", override = true }

[addresses]
myproj = "0x0"

# Named addresses will be accessible in Move as `@name`. They're also exported:
# for example, `std = "0x1"` is exported by the Standard Library.
# alice = "0xA11CE"

[dev-dependencies]
# The dev-dependencies section allows overriding dependencies for `--test` and
# `--dev` modes. You can introduce test-only dependencies here.
# Local = { local = "../path/to/dev-build" }

[dev-addresses]
# The dev-addresses section allows overwriting named addresses for the `--test`
# and `--dev` modes.
# alice = "0xB0B"

==== myproj/.gitignore ====
build/*
==== myproj/sources/myproj.move ====
/*
/// Module: myproj
module myproj::myproj;
*/

// For Move coding conventions, see
// https://docs.sui.io/concepts/sui-move-concepts/conventions


==== myproj/tests/myproj_tests.move ====
/*
#[test_only]
module myproj::myproj_tests;
// uncomment this line to import the module
// use myproj::myproj;

const ENotImplemented: u64 = 0;

#[test]
fun test_myproj() {
    // pass
}

#[test, expected_failure(abort_code = ::myproj::myproj_tests::ENotImplemented)]
fun test_myproj_fail() {
    abort ENotImplemented
}
*/

----- stderr -----
